Cookbook/Proposals/Base class for FeedbackLoops systems.md

```diff
@@ -0,0 +1,157 @@
+# Base class for FeedbackLoops systems
+
+* Author(s): Sergey Shulga
+* Review Manager: Anders Ha
+
+## Introduction
+
+Almost for all our screens, we implement ViewModels which are powered by the [ReactiveFeeedback](https://github.com/Babylonpartners/ReactiveFeedback).
+
+The current interface of VM looks like this
+
+```swift
+public protocol BoxViewModel: AnyObject {
+    associatedtype State
+    associatedtype Action
+
+    var state: Property<State> { get }
+
+    func send(action: Action)
+}
+```
+
+## Motivation
+
+One of the main issues with this approach
+
+1. We need to implement `input` feedback in pretty much every VM which may lead to errors if do not schedule input correctly, for this we have `Feedback.input()` convenience extension
+
+2. If the screen contains a lot of input field we end up having a lot of event's that are just mutate one property of the `State` e.g
+`Action.didChangeUserName(String)` `Action.didPickBirthDate(Date)`
+ 
+## Proposed solution
+1. Implementation of the feedback:
+
+```swift
+open class FeedbackLoop<State, Event, Action>: FeedbackLoopType {
+  public let state: Property<State>
+  private let input = Feedback<State, Update>.input()
+  private let actionTransformer: (Action) -> Event
+
+  public init(
+    initial: State,
+    reducer: @escaping (State, Event) -> State,
+    feedbacks: [Feedback<State, Event>],
+    scheduler: DateScheduler,
+    actionTransformer: @escaping (Action) -> Event
+  ) {
+    self.actionTransformer = actionTransformer
+    self.state = Property(
+      initial: initial,
+      scheduler: scheduler,
+      reduce: { (state: State, update: Update) -> State in
+        switch update {
+        case let .event(event):
+          return reducer(state, event)
+        case let .mutation(mutation):
+          return mutation.mutate(state)
+        }
+      },
```

Comment:
This feels like a higher order reducer.

```
func mutationReducer<State, Event>(_ reducer: (State, Event) -> State) -> (State, Event) -> State {
    return { state, event in 
       switch event {
              case let .event(anEvent):
                     return reducer(state, anEvent)
              case let .mutation(mutation):
                     return mutation.mutate(state)
              }
    }
}
```

And then:

```swift
self.state = Property(
	      initial: initial,
	      scheduler: scheduler,
	      reduce: mutationReducer(reducer),
              feedbacks: feedbacks.map { $0.mapEvent(Update.event) } 
```

Reply:
Yes, it's a lift from `Event` to `Update`:

```swift
func mutationReducer<State, Event>(_ reducer: (State, Event) -> State) 
    -> (State, Update) -> State
```

But what `(State, Update) -> State` basically means is that we will have a **direct way to mutating sub-state at anytime**, which is against the concept of (well-designed) state machine that has the minimal state transition arrows.
If we allow `Update` to be passed in, it means we are going to draw arbitrary arrows between two states, which will be just like **mutable state `class` with `var`s**.

Reply:
Ah yes, I missed the state lifting from the reducer passed in the init, to the reducer passed to the state machine.

Reply:
> If we allow `Update` to be passed in, it means we are going to draw arbitrary arrows between two states, which will be just like **mutable state `class` with `var`s**.

That's exactly my concern too.

Reply:
> mutable state class with vars

That's not entirely correct, because mutation will emit new State

Reply:
Well I guess it's a trade off of dropping some actions that do not give much value. To some extent we just bind a component to particular sub state if only thing it does it updating its value ¯\_(ツ)_/¯

Reply:
But if the emitting event does something else e.g triggering another feedback or does some validation of input than we still are going to do it via `Action.someChange(String)`.

Reply:
@sergdort is it the goal to update the state via the KeyPath conveniences?

Reply:
can we address @mluisbrown and @inamiy concern by leaving handling of mutation in each view model rather than doing that in the base class? essentially mutation is a pair of key path and value, so why not to bundle it in one generic even? this seems to be the main idea of the suggestion, isn't it?

Reply:
It also will hurt debugging if handling of all mutations is done in one central place, wouldn't it?

Reply:
My two cents for this solution is to add **`SubState`** instead, and **disallow `Mutation` that receives current state as input** (since current-state referencing i.e. `Reducer` is a complicated mechanism that should only be handled by state machine, not each caller's sending).

```swift
struct State {
    var subState: SubState // Mutation-aware
    var others: ...  // not Mutation-aware
    
    // Explicitly define sub-state for mutation.
    struct SubState {
    	...
    }
}

enum Update {
    case event(Event)
    
    // NOTE: Directly passing a new sub-state WITHOUT referring current sub-state.
    // Behaviour is mostly the same as SwiftUI's direct state binding.
    case syncSubState(SubState)
}
```

This literally means **`SubState` will also behave as sub-action**,
which then allows us to not draw arbitrary state transition arrows, so it's safer.

Reply:
Ah pls forget above.
`SubState` won't be fully able to behave as sub-action, since struct and enum are not the same but dual.

Reply:
@ilyapuchka 
> It also will hurt debugging if handling of all mutations is done in one central place, wouldn't it?

You would need to add it in every view model

Reply:
but all the mutations are handled here in the base class, no code to handle key path mutations would be in the view models, which seems to be the point of proposal. so considering debugging the only point where we will be able to set breakpoint will be this switch and it will hit it on all view models, there will be no way to debug a single view model conveniently.